pairing/plonk/src/prover.rs

```diff
@@ -0,0 +1,293 @@
+
+This source file implements prover's zk-proof primitive.
+
+*********************************************************************************************/
+
+use rand_core::OsRng;
+use algebra::{Field, PairingEngine};
+use oracle::rndoracle::{ProofError};
+use ff_fft::{DensePolynomial, Evaluations};
+use commitment_pairing::commitment::Utils;
+pub use super::index::Index;
+use oracle::sponge::FqSponge;
+use crate::plonk_sponge::FrSponge;
+
+#[derive(Clone)]
+pub struct ProverProof<E: PairingEngine>
+{
+    // polynomial commitments
+    pub a_comm: E::G1Affine,
+    pub b_comm: E::G1Affine,
+    pub c_comm: E::G1Affine,
+    pub z_comm: E::G1Affine,
+    pub tlow_comm: E::G1Affine,
+    pub tmid_comm: E::G1Affine,
+    pub thgh_comm: E::G1Affine,
+
+    // batched commitment opening proofs
+    pub proof1: E::G1Affine,
+    pub proof2: E::G1Affine,
+
+    // polynomial evaluations
+    pub evals : ProofEvaluations<E::Fr>,
+
+    // public part of the witness
+    pub public: Vec<E::Fr>
+}
+
+impl<E: PairingEngine> ProverProof<E>
+{
+    // This function constructs prover's zk-proof from the witness & the Index against URS instance
+    //     witness: computation witness
+    //     index: Index
+    //     RETURN: prover's zk-proof
+    pub fn create
+        <EFqSponge: FqSponge<E::Fq, E::G1Affine, E::Fr>,
+         EFrSponge: FrSponge<E::Fr>,
+        >
+    (
+        witness: &Vec::<E::Fr>,
+        index: &Index<E>
+    ) -> Result<Self, ProofError>
+    {
+        let n = index.cs.domain.size();
+        if witness.len() != 3*n {return Err(ProofError::WitnessCsInconsistent)}
+
+        let mut oracles = RandomOracles::<E::Fr>::zero();
+        let mut evals = ProofEvaluations::<E::Fr>::zero();
+
+        // the transcript of the random oracle non-interactive argument
+        let mut fq_sponge = EFqSponge::new(index.fq_sponge_params.clone());
+
+        // compute public input polynomial
+        let public = witness[0..index.cs.public].to_vec();
+        let p = -Evaluations::<E::Fr>::from_vec_and_domain(public.clone(), index.cs.domain).interpolate();
+
+        let a = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.l.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+        let b = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.r.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+        let c = &Evaluations::<E::Fr>::from_vec_and_domain(index.cs.gates.iter().map(|gate| witness[gate.o.0]).collect(), index.cs.domain).interpolate()
+            + &DensePolynomial::rand(1, &mut OsRng).mul_by_vanishing_poly(index.cs.domain);
+
+        // commit to the a, b, c wire values
+        let a_comm = index.urs.get_ref().commit(&a)?;
```

Comment:
Maybe we can call them either `a, b, c` or `l, r, o` everywhere for consistency?

Reply:
Agreed, will do calling them all l, r, o